src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -499,46 +506,92 @@ newIncomingMergingRun tr hfs hbio activeDir uc
      TableConfig
   -> LevelNo
   -> IncomingRun IO h
-  -> MergeCredits
+  -> NominalCredits
   -> IO () #-}
--- | Supply a given number of credits to the merge in an incoming run.
+-- | Set the  a given number of credits to the merge in an incoming run.
 supplyCreditsIncomingRun ::
      (MonadSTM m, MonadST m, MonadMVar m, MonadMask m)
   => TableConfig
   -> LevelNo
   -> IncomingRun m h
-  -> MergeCredits
+  -> NominalCredits
   -> m ()
-supplyCreditsIncomingRun _ _     (Single        _r) _       = return ()
-supplyCreditsIncomingRun conf ln (Merging _ _ _ mr) credits = do
-    let !thresh = creditThresholdForLevel conf ln
-    _ <- MR.supplyCreditsRelative mr thresh credits
+supplyCreditsIncomingRun _ _ (Single _r) _ = return ()
+supplyCreditsIncomingRun conf ln (Merging _ nominalDebt nominalCreditsVar mr)
+                         deposit = do
+    (_nominalCredits,
+     nominalCredits') <- depositNominalCredits nominalDebt nominalCreditsVar
+                                               deposit
+    let !mergeDebt     = MR.totalMergeDebt mr
+        !mergeCredits' = scaleNominalToMergeCredit nominalDebt mergeDebt
+                                                   nominalCredits'
+        !thresh = creditThresholdForLevel conf ln
+    (_suppliedCredits,
+     _suppliedCredits') <- MR.supplyCreditsAbsolute mr thresh mergeCredits'
+    --TODO: consider tracing supply of credits
     return ()
 
+depositNominalCredits ::
+     PrimMonad m
+  => NominalDebt
+  -> PrimVar (PrimState m) NominalCredits
+  -> NominalCredits
+  -> m (NominalCredits, NominalCredits)
+depositNominalCredits (NominalDebt nominalDebt)
+                      nominalCreditsVar
+                      (NominalCredits deposit) = do
+    NominalCredits before <- readPrimVar nominalCreditsVar
+    -- Depositing _could_ leave the credit higher than the debt, because
+    -- sometimes under-full runs mean we don't shuffle runs down the levels
+    -- as quickly as the worst case. So here we do just drop excess nominal
+    -- credits.
```

Comment:
I probably should have commented this on the prototype, but the scenario described in this comment is not completely clear to me. Am I interpreting it correctly: when a completed merge is underfull, we keep it on the same level, so it means we don't call `expectCompleted` on any levels further down the tree, which means we keep supplying nominal credits to levels further down, and that means we can go over the nominal debt?

Reply:
Exactly.

Reply:
I'll improve the comment here.